File: storage/device.rb
require_relative 'device_and_reading_storage'
require_relative 'readings_association'


class Device
  attr_reader :id, :uid


  def initialize(uid:)
    @id = nil
    @uid = uid
  end


  def save!
    unless @id
      storage = DeviceAndReadingStorage.instance
      if storage.find_device_by_uid(@uid)
        raise "Device with uid '#{@uid}' already exists"
      end
      storage.add_device(self)
    end
    self
  end


  def persisted?
    !@id.nil?
  end


  def readings
    raise "Device must be persisted before accessing readings" unless persisted?
    @readings_association ||= ReadingsAssociation.new(@id)
  end


  def total_count
    readings.sum(&:count)
  end


  def latest_timestamp
    readings.map(&:timestamp).max
  end


  class << self

    def find(id)
      DeviceAndReadingStorage.instance.find_device_by_id(id)
    end


    def find_by_uid(uid)
      DeviceAndReadingStorage.instance.find_device_by_uid(uid)
    end


    def create!(uid:)
      device = new(uid: uid)
      device.save!
      device
    end

  end

end

File: express_api/src/device.js
const storage = require('./storage');
const Reading = require('./reading');


class Device {

  constructor(uid) {
    this.id = null;
    this.uid = uid;
  }


  save() {
    if (!this.id) {
      if (storage.findDeviceByUid(this.uid)) {
        throw new Error(`Device with uid '${this.uid}' already exists`);
      }
      storage.addDevice(this);
    }
    return this;
  }


  getTotalCount() {
    return storage.getDeviceTotalCount(this.id);
  }


  getLatestTimestamp() {
    return storage.getDeviceLatestTimestamp(this.id);
  }


  findOrCreateReading(timestamp, count) {
    const existingReading = storage.findReadingByDeviceAndTimestamp(this.id, timestamp);
    if (existingReading) {
      return existingReading;
    }
    const reading = new Reading(this.id, timestamp, count);
    reading.save();
    return reading;
  }


  static findByUid(uid) {
    return storage.findDeviceByUid(uid);
  }


  static create(uid) {
    const device = new Device(uid);
    device.save();
    return device;
  }


  static findOrCreate(uid) {
    return storage.findDeviceByUid(uid) || Device.create(uid);
  }

}


module.exports = Device;
